source/common/router/router.cc

```diff
@@ -997,13 +1004,36 @@ void Filter::resetOtherUpstreams(UpstreamRequest& upstream_request) {
   final_upstream_request->moveIntoList(std::move(final_upstream_request), upstream_requests_);
 }
 
+void Filter::maybeSetGrpcStatusToOutlierDetection(
+    UpstreamRequest& upstream_request, absl::optional<Grpc::Status::GrpcStatus> grpc_status) {
+  if (grpc_status) {
+    if (Http::CodeUtility::is5xx(Grpc::Utility::grpcToHttpStatus(grpc_status.value()))) {
+      // For gRPC requests, we overwrite the default (200) status for the outlier detector
+      // only when the grpc status maps to 5xx HTTP status.
+      // TODO: confirm if 4xx status should be recorded with the outlier detector.
+      upstream_request.upstream_host_->outlierDetector().putHttpResponseCode(
+          Grpc::Utility::grpcToHttpStatus(grpc_status.value()));
+    } else {
+      upstream_request.upstream_host_->outlierDetector().putHttpResponseCode(
+          http_status_from_grpc_response_);
+    }
+    http_status_from_grpc_response_ = 0;
+  }
```

Comment:
This reset looks questionable ... and we should only do putResonseCode() to when the stream is finished or reset